packages/flutter/lib/src/widgets/async.dart

```diff
@@ -0,0 +1,387 @@
+  ///
+  /// Sub-classes must override this method to provide the initial value for
+  /// the fold computation.
+  S initial();
+
+  /// Returns an updated version of the [current] summary reflecting that we
+  /// are now connected to a stream.
+  ///
+  /// The default implementation returns [current] as is.
+  S afterConnected(S current) => current;
+
+  /// Returns an updated version of the [current] summary following a data event.
+  ///
+  /// Sub-classes must override this method to specify how the current summary
+  /// is combined with the new data item in the fold computation.
+  S afterData(S current, T data);
+
+  /// Returns an updated version of the [current] summary following an error.
+  ///
+  /// The default implementation returns [current] as is.
+  S afterError(S current, Object error) => current;
+
+  /// Returns an updated version of the [current] summary following stream
+  /// termination.
+  ///
+  /// The default implementation returns [current] as is.
+  S afterDone(S current) => current;
+
+  /// Returns an updated version of the [current] summary reflecting that we
+  /// are no longer connected to a stream.
+  ///
+  /// The default implementation returns [current] as is.
+  S afterDisconnected(S current) => current;
+
+  /// Returns a Widget based on the [currentSummary].
+  Widget build(BuildContext context, S currentSummary);
+
+  @override
+  State<StreamBuilderBase<T, S>> createState() => new _StreamBuilderBaseState<T, S>();
+}
+
+/// State for [StreamBuilderBase].
+class _StreamBuilderBaseState<T, S> extends State<StreamBuilderBase<T, S>> {
+  StreamSubscription<T> _subscription;
+  S _summary;
+
+  @override
+  void initState() {
+    super.initState();
+    _summary = config.initial();
+    _subscribe();
+  }
+
+  @override
+  void didUpdateConfig(StreamBuilderBase<T, S> oldConfig) {
+    if (oldConfig.stream != config.stream) {
+      if (_subscription != null) {
+        _unsubscribe();
+        _summary = config.afterDisconnected(_summary);
+      }
+      _subscribe();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) => config.build(context, _summary);
+
+  @override
+  void dispose() {
+    _unsubscribe();
+    super.dispose();
+  }
+
+  void _subscribe() {
+    if (config.stream != null) {
+      _subscription = config.stream.listen((T data) {
+        setState(() {
+          _summary = config.afterData(_summary, data);
+        });
+      }, onError: (Object error) {
+        setState(() {
+          _summary = config.afterError(_summary, error);
+        });
+      }, onDone: () {
+        setState(() {
+          _summary = config.afterDone(_summary);
+        });
+      });
+      _summary = config.afterConnected(_summary);
+    }
+  }
+
+  void _unsubscribe() {
+    if (_subscription != null) {
+      _subscription.cancel();
+      _subscription = null;
+    }
+  }
+}
+
+/// The state of connection to an asynchronous computation.
+///
+/// See also:
+///
+/// * [AsyncSnapshot]
```

Comment:
nit: here and below, when you have a bullet list of classes in a "see also" section, please add a comment describing why that class is interesting, as in:
```markdown
/// * [AsyncSnapshot], which describes foo bar baz.
```